### DISCUSSION TITLE
OM on linux: missing standard library...

### DISCUSSION THREAD

**Discussion by Rhinou** [OpenModelica/OpenModelica#12791] [General]
I just installed OM on linux using instructions on the website, but when I start OMEdit, it does not load any library. In "Messages" window there is:

```
[1] 22:29:02 Scripting Notification
You can install the requested package using one of the commands:
  installPackage(Modelica, "4.0.0+maint.om", exactMatch=true)
  installPackage(Modelica).

[2] 22:29:02 Scripting Error
Failed to load package Modelica (default) using MODELICAPATH /home/user/.openmodelica/libraries/.

[3] 22:29:33 Scripting Notification
[/home/user/.openmodelica/libraries/index.json: 0:0-0:0]: Downloaded package index from URL https://libraries.openmodelica.org/index/v1/index.json.
```

I tried to install library with "sudo apt install omlibrary", but still nothing. Only messages changed to:

```
[1] 22:39:23 Translation Notification
[/usr/bin/../share/omlibrary/cache/index.json: 0:0-0:0]: Cached libraries were found and will be installed into /home/user/.openmodelica/libraries/.

[2] 22:39:23 Error
Failed to open file: /home/user/.openmodelica/cache/96032134c36668898e1693e69bd9f81aa38de3dd.zip

[3] 22:39:23 Scripting Notification
After extracting /home/user/.openmodelica/cache/96032134c36668898e1693e69bd9f81aa38de3dd.zip, /home/user/.openmodelica/libraries/Complex 4.0.0+maint.om/package.mo does not exist. Removing the failed installation.

[4] 22:39:23 Translation Error
Class Modelica not found.
```

How can I fix this?

**Comment by casella** [comment]
You should type those commands in the OMShell console.

**Reply by Rhinou** [reply]
Thank you, this worked, but... why does it have to be this complicated? 

I installed OM 1.23.1 on Windows, and when I start OMEdit there, all main libraries (OpenModelica, ModelicaServices, Complex, Modelica) were loaded automatically. No manual intervention was necessary, no OMShell scripting. It just worked, as user is expecting.

Why it must be more complicated for Linux? To discourage users from using it? Imagine common Windows-users, trying Linux for the first time: been using OM on Windows for years and never had to open OMShell (and let's be honest: most of Windows-users never used any shell, console, terminal, etc). If on Linux he must write some cryptic commands in some console just to get the same functionality, he will turn around and goes back to windows saying, "Linux is for nerds, it forces users to use console and type some command you do not even know what it does, just to get what we have comfortably on much more user-friendly Windows".

I'm pleased when more and more software is being ported to Linux. But the way OpenModelica4Linux is made, it might be doing more harm than good...

**Comment by casella** [comment]
@Rhinou as far as I am concerned, I agree with you.

@sjoelund, @adrpo, I remember we had this discussion in the past, but I can't remember the details. Why can we have the MSL installed by default also on Linux?

**Reply by sjoelund** [reply]
OMEdit has a `suggests` dependency on the omlibrary package. We could change that to a `recommends`, but if using the nightly... That means updating the library cache all the time even though it will only be used on first start of the application... The cache is larger than all other OpenModelica packages combined.

**Reply by casella** [reply]
@sjoelund does omlibrary only contain the MSL and related packages?

**Reply by sjoelund** [reply]
Only MSL 3.2.3 and 4.0.0

**Reply by AnHeuermann** [reply]
Could we add a popup to OMEdit when starting it for the first time that asks the user to install MSL 3.2.3 and MSL 4.0.0?
```
You started OMEdit for the first time. Shall OMEdit download and install the Modelica standard libraries?

Yes / Ask me later / No, don't ask me again
```

I guess it would be a tiny script
```modelica
updatePackageIndex();
installPackage(Modelica, ""); // install latest MSL version
installPackage(Modelica, "4.0.0+maint.om", exactMatch=true); // install MSL 4.0.0
installPackage(Modelica, "3.2.3+maint.om", exactMatch=true); // install MSL 3.2.3
```
that @adeas31 could start from that GUI dialog window. Maybe we add a check if the libraries are already installed before opening the dialog.

**Reply by casella** [reply]
@AnHeuermann the reason why we pre-installed MSL in Windows is to support the case of installations that cannot access the internet for some reason (e.g. behind a firewall, or large classrooms without enough bandwidth to the outside network). That requirement still holds. Unfortunately it was not possible to do exactly the same on Linux.

On this point, I am still wondering if we could not actually do it the same way we do it on Windows, without involving the Linux package managers. The point is to have a cache with basic MSL versions that the package manager can access without the need of going on the net. @sjoelund can't we just do that?

> OMEdit has a suggests dependency on the omlibrary package. We could change that to a recommends, but if using the nightly... That means updating the library cache all the time even though it will only be used on first start of the application... The cache is larger than all other OpenModelica packages combined.

Note that this cache is really not meant to be updated, the idea is to use the OMC library package manager to do that. So I really see no strong reason to use the Linux package manager for that. It's just a bunch of files which is part of the regular OMC installation.

**Reply by sjoelund** [reply]
> The point is to have a cache with basic MSL versions that the package manager can access without the need of going on the net. @sjoelund can't we just do that?

We do that if you install the omlibrary package. But it's then updated, which is silly. And you need an internet connection to install the package to begin with, so it would make more sense for OMEdit to just install the packages from the internet if it's not there.

Perhaps for the virtual machine it could be installed and then `apt hold omlibrary` and the instructions for the webpage could do the same.

**Reply by casella** [reply]
> We do that if you install the omlibrary package. But it's then updated, which is silly. 

Yes, that is the problem.

> And you need an internet connection to install the package to begin with, so it would make more sense for OMEdit to just install the packages from the internet if it's not there.

> Perhaps for the virtual machine it could be installed and then `apt hold omlibrary` and the instructions for the webpage could do the same.

Good idea. @sjoelund could you please take care of that (or delegate it)?

**Reply by casella** [reply]
Or maybe @adrpo?
